pkg/network/go/bininspect/newproc.go

```diff
@@ -0,0 +1,252 @@
+	//   pointer at %fs-8 on 64 bit.
+	// - %Gs is the index of private storage in GDT on 32 bit, and puts the G
+	//   pointer at -4(tls).
+	// - Otherwise, Go asks the external linker to place the G pointer by
+	//   emitting runtime.tlsg, a TLS symbol, which is relocated to the chosen
+	//   offset in libc's TLS block.
+	// - On ARM64 (but really, any architecture other than i386 and 86x64) the
+	//   offset is calculated using runtime.tls_g and the formula is different.
+
+	var tls *elf.Prog
+	for _, prog := range i.elf.file.Progs {
+		if prog.Type == elf.PT_TLS {
+			tls = prog
+			break
+		}
+	}
+
+	switch i.elf.arch {
+	case GoArchX86_64:
+		tlsg, ok := i.symbols["runtime.tlsg"]
+		if !ok || tls == nil {
+			return ^uint64(i.elf.arch.PointerSize()) + 1, nil //-ptrSize
+		}
+
+		// According to https://reviews.llvm.org/D61824, linkers must pad the actual
+		// size of the TLS segment to ensure that (tlsoffset%align) == (vaddr%align).
+		// This formula, copied from the lld code, matches that.
+		// https://github.com/llvm-mirror/lld/blob/9aef969544981d76bea8e4d1961d3a6980980ef9/ELF/InputSection.cpp#L643
+		memsz := tls.Memsz + (-tls.Vaddr-tls.Memsz)&(tls.Align-1)
+
+		// The TLS register points to the end of the TLS block, which is
+		// tls.Memsz long. runtime.tlsg is an offset from the beginning of that block.
+		return ^(memsz) + 1 + tlsg.Value, nil // -tls.Memsz + tlsg.Value
+
+	case GoArchARM64:
+		tlsg, ok := i.symbols["runtime"]
+		if !ok || tls == nil {
+			return 2 * uint64(i.elf.arch.PointerSize()), nil
+		}
+
+		return tlsg.Value + uint64(i.elf.arch.PointerSize()*2) + ((tls.Vaddr - uint64(i.elf.arch.PointerSize()*2)) & (tls.Align - 1)), nil
+
+	default:
+		return 0, fmt.Errorf("binary is for unsupported architecture")
+	}
+}
+
+// getGoroutineIDMetadata collects enough metadata about the binary
+// to be able to reliably determine the goroutine ID from the context of an eBPF uprobe.
+// This is accomplished by finding the offset of the `goid` field in the `runtime.g` struct,
+// which is the goroutine context struct.
+//
+// A pointer to this struct is always stored in thread-local-strorage (TLS),
+// but it might also be in a dedicated register (which is faster to access),
+// depending on the ABI and architecture:
+// 1. If it has a dedicated register, this function gives the register number
+// 2. Otherwise, this function finds the offset in TLS that the pointer exists at.
+//
+// See:
+// - https://go.googlesource.com/proposal/+/master/design/40724-register-calling.md#go_s-current-stack_based-abi
+// - https://go.googlesource.com/go/+/refs/heads/dev.regabi/src/cmd/compile/internal-abi.md#amd64-architecture
+// - https://github.com/golang/go/blob/61011de1af0bc6ab286c4722632719d3da2cf746/src/runtime/runtime2.go#L403
+// - https://github.com/golang/go/blob/61011de1af0bc6ab286c4722632719d3da2cf746/src/runtime/runtime2.go#L436
+func (i *newProcessBinaryInspector) getGoroutineIDMetadata(abi GoABI) (GoroutineIDMetadata, error) {
```

Comment:
Consider add UTs to ensure the function works correctly over different architecture
if that's too complex, please add jira ticket to handle it later

Reply:
I'll add a ticket for now and we'll address the tests as a separate task for the inspection